http_service/tests/test_readthrough_cache.py

```diff
@@ -0,0 +1,125 @@
+    mockdatetime = MockDatetime(datetime(2019, 4, 1, 10))
+    cache = ReadthroughTTLCache(timedelta(hours=4), lambda x: "payload")
+
+    with patch("datetime.datetime", mockdatetime):
+        cache.get("key_a")
+
+        # after one hour
+        mockdatetime.set_now(datetime(2019, 4, 1, 11))
+        assert "key_a" not in cache
+
+        # after two hours
+        mockdatetime.set_now(datetime(2019, 4, 1, 12))
+        cache.get("key_a")
+
+        # after three hours
+        mockdatetime.set_now(datetime(2019, 4, 1, 13))
+        assert "key_a" in cache
+
+
+def test_cache_purges_after_ttl():
+    mockdatetime = MockDatetime(datetime(2019, 4, 1, 10))
+    cache = ReadthroughTTLCache(timedelta(hours=2), lambda x: "payload")
+
+    with patch("datetime.datetime", mockdatetime):
+        cache.get("key_a", force_store=True)
+
+        # after one hour
+        mockdatetime.set_now(datetime(2019, 4, 1, 11))
+        cache.purge_expired_entries()
+        assert "key_a" in cache
+
+        # after two hours one minute
+        mockdatetime.set_now(datetime(2019, 4, 1, 12, 1))
+        cache.purge_expired_entries()
+        assert "key_a" not in cache
+
+
+def test_cache_ttl_refreshes_after_get():
+    mockdatetime = MockDatetime(datetime(2019, 4, 1, 10))
+    cache = ReadthroughTTLCache(timedelta(hours=2), lambda x: "payload")
+
+    with patch("datetime.datetime", mockdatetime):
+        cache.get("key_a", force_store=True)
+
+        # after one hour
+        mockdatetime.set_now(datetime(2019, 4, 1, 11))
+        cache.purge_expired_entries()
+        assert "key_a" in cache
+        assert cache.get("key_a") == "payload"
+
+        # after three hours
+        mockdatetime.set_now(datetime(2019, 4, 1, 13))
+        cache.purge_expired_entries()
+        assert "key_a" in cache
+
+        # after three hours one minute
+        mockdatetime.set_now(datetime(2019, 4, 1, 13, 1))
+        cache.purge_expired_entries()
+        assert "key_a" not in cache
+
+
+def test_force_store():
+    cache = ReadthroughTTLCache(timedelta(hours=2), lambda x: "payload")
+    cache.get("key_a", force_store=True)
+    assert "key_a" in cache
+
+
+def test_cache_thread():
+    mockdatetime = MockDatetime(datetime(2019, 4, 1, 10))
+    cache = ReadthroughTTLCache(timedelta(hours=2), lambda x: "payload")
+    mocksleep = MockSleep(mockdatetime)
+    with patch("datetime.datetime", mockdatetime):
+        with patch("time.sleep", mocksleep.sleep):
+            cache.get("key_a", force_store=True)
+            cache.start_ttl_thread()
+
+            # after one hour
+            mockdatetime.set_now(datetime(2019, 4, 1, 11))
+            assert "key_a" in cache
+            assert mocksleep.wakeups_count == 0
+
+            # after two hours one minute
+            before_purge_timestamp = cache.last_purged_time
+            mockdatetime.set_now(datetime(2019, 4, 1, 12, 1))
+            while cache.last_purged_time == before_purge_timestamp:
```

Comment:
You could instead mock purge_expired_entries to wait until it is called, this way you can also remove the `last_purged_time` attribute.

Reply:
(and while waiting, you could use `time.sleep(0)` instead of `pass`, to yield control to the cleanup thread if needed and avoid possible hangs)